### PR TITLE
Diffie-hellman implementation

### PR DESCRIPTION
**Pull request by crackcomm**
Includes small test/benchmark/demo to use with `python3 -m http.server`.

### CODE REVIEW

**Comment by lights0123** [comment]
This adds 10kB (68kB → 77kB) to the wasm file, which is probably acceptable. re: import maps: do they work? Chrome has them behind a feature flag, so I think it would be best to just spell out `./pkg/wasm_x25519.js`.

File: index.html [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+      }
+    }
+  </script>
+</head>
+
+<body id="body"></body>
+<script type="module">
+  import { generate_keypair, diffie_hellman, default as init } from 'wasm_x25519';
+
+  init('./pkg/wasm_x25519_bg.wasm').then(() => {
```
Comment (lights0123) [inline index.html]:
```suggestion
  init().then(() => {
```

not needed

File: index.html [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+      "imports": {
+        "wasm_x25519": "./pkg/wasm_x25519.js"
+      }
+    }
+  </script>
+</head>
+
+<body id="body"></body>
+<script type="module">
+  import { generate_keypair, diffie_hellman, default as init } from 'wasm_x25519';
```
Comment (lights0123) [inline index.html]:
```suggestion
  import { generate_keypair, diffie_hellman, default as init } from './pkg/wasm_x25519.js';
```

File: index.html [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+<head>
+  <meta charset="utf-8" />
+  <script type="importmap">
+    {
+      "imports": {
+        "wasm_x25519": "./pkg/wasm_x25519.js"
+      }
+    }
+  </script>
```
Comment (lights0123) [inline index.html]:
```suggestion
```

File: src/lib.rs [hunk 4]
```diff
@@ -20,16 +21,32 @@ fn b64_encode<'a>(input: &[u8], out: &'a mut [u8]) -> &'a mut [u8] {
...
+    base64::decode_config_slice(input, base64::STANDARD, out)
+}
+
+fn encode_key(secret: &[u8]) -> JsString {
+    let mut encoded = [0; KEY_SIZE_BASE64];
+    from_utf8(b64_encode(secret, &mut encoded)).unwrap().into()
+}
+
+#[wasm_bindgen(catch)]
+pub fn diffie_hellman(secret_key: &str, public_key: &str) -> Result<JsString, JsValue> {
```
Comment (lights0123) [inline src/lib.rs:35]:
```suggestion
/// Calculate a shared secret from one party's secret key and another party's public key.
pub fn diffie_hellman(secret_key: &str, public_key: &str) -> Result<JsString, JsValue> {
```

File: index.html [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+      const [secret2, public2] = generate_keypair();
+      const shared1 = diffie_hellman(secret2, public1);
+      const shared2 = diffie_hellman(secret1, public2);
+
+      console.log(public1, public2, shared1 === shared2, new Date() - start);
+    }
+  });
+</script>
+
+</html>
```
Comment (lights0123) [inline index.html]:
```suggestion
</html>

```

**Review by lights0123 — CHANGES_REQUESTED** [review]

**Comment by crackcomm** [comment]
I compared this library to x25519 in bcrypto, it's still smaller and faster.

**Comment by lights0123** [comment]
Thanks! Released as 0.1.1 on npm

**Pull request merged**